### PR TITLE
Simplify caching mechanisms for CI and PROD images (regular PR)

### LINKED ISSUE #42999: "Main" tip caching of CI image is broken
It looks like after the provider move in #42505, "main" tip caching for airflow is broken. Before the change installing airflow had use two stages caching:

1) Airflow has been installed from main first (which allowed to perform pre-installation without invalidation of docker layer)
2) Then additional installation was done after copying latest sources and pyproject.toml/hatch

This allowed for very fast rebuilds of the image using remote cache built on ghcr.io, when only source files or dependencies of airlfow changed.

It usually took 2-5 seconds to rebuild such image. Currently it looks like second installation step is running again full installation of airlfow (it takes about 2 minutes), which means that likely some mechanism of caching is broken.

Image

We should review and updated Dockerfile.ci and scripts to bring the speed back.

### LINKED ISSUE #43268: Remove "pull_request_target" workflows and replace them with storing images as artifacts
I've learned that as of artifact@v4 APIs for GitHub Actions, it should be possible to get rid of the GitHub Registry use as CI cache.

> [!NOTE]
> I think that change might be way smaller. We should be able to continue using Github Registry for both - latest image and image cache, and only change CI to pull artifacts from latest "canary" build and build image cache from that.

In our CI/CD pipelines we are using "pull_request_target" workflow in order to be able to publish images to GitHub registry from Pull Requests coming from forks. This is done in order to speed up the workflows - our images take a lot of time to build and by reusing the cache, we are saving a lot of build time.

However after following a discussion in Apache Software Foundation `#builds` channel https://the-asf.slack.com/archives/CF6PY5M4N/p1729595424091089 - with Pulsar and Netbeans CI/CD maintainers, it looks like we have a viable alternative - that will remove the need of using GH Registry for PRs - instead we could use artifacts.

Our images are quite big (2GB) but Pulsar team successfully uses artifacts to store 1.7GB ones, so this should be quite possible for us to use the artifacts instead of registry.

We dicussed also - with Pulsar and Netbeans CI maintainers about caching strategy we use and it seems entirely doable to switch to artifacts and get the same effect, without the need of having "pull_request_target" workflow. That would be a major win for security of our workflows to switch to it.

This is especially atractive because the new artifact API v4 introduced in February 2024 is very fast, implements automated chunking of binaries uploaded/downloaded and parallell uploads/downloads - which according to Pulsar/Netbeans team had a huge impact on upload/download times for the images - see https://github.blog/news-insights/product-news/get-started-with-v4-of-github-actions-artifacts/ - there are a lot of details in the post, but the way uploading artifacts works with v4 is that you directly interact with Azure Blob Storage, the files are chunked etc. The examples they show in the blog post is speeding up the upload of 1GB file from 2 m to 3 s.

Also v4 when `run-id` and other parameters have been added - making it suitable to implement similar caching solution we have, where artifacts from "canary" build could be used by all PRs as source of their image cache.

Seems that artifacts have no individual size limits and they also do not have "total limits" with the ASF "Enterprise Account" level - there are things 

The way it **should** work:

All workflows - PRs from apache and fork work in the same way:

1) Pull latest image (stored as artifact) from latest "main" successful build (we can use GH API to find the run-id to pull it from and the download-artifact@v4 action already supports pulling artifacts from other workflows and from other runs.

2) The PRs first unpack and push the "main" image to be present in local docker registry - no need to have GH Registry "PUSH" access.

3) The PRs then build the image locally not using `--cache-from` but to use such local registry cache. Achieving basically the same (or even bettter) caching speedups comparing to what we have now.

4) The successful "canary" build artifact could be uploaded at "finalize" time to become the new source of cache for all the PRs.

5) Breeze locak CI  image could also be modified to pull latest image when rebuilding the image forcefully - otherwise rely on locally built images. 

6) The `--image-tag` option of the local breeze (which pulls the image from a specific workflow run) could be easily adapted to be changed to `--pr PR#` - allowing to replicate a failure from specific PR locally without looking up the image.

7) The artifacts can have arbitrary set expiry time, so they could be available for a few days after the PR or "canary" build completed.

The benefits:

* we get rid of "Build image" "pull_request_target" workflow - that would be a huge security improvement
* we get rid of the need of creating new images every time we create a new "release" branch.
* we get rid of GitHub Registry
* all the runs - commiter and non-commiter, from fork or apache would be identical, making it much easier to test changes in workflows
* looking at the upload/download stats for artifacts, it is likely that switching to artifacts might get some speedup comparing to using the registry

We can use https://github.com/assignUser/stash from Jacob Wujciak (ASF member) to make things easier. This will soon be available as Apache-wide action: https://github.com/apache/infrastructure-actions/pull/32

### PR DESCRIPTION
**Pull request by potiuk**
For a long time we had used a sophisticated mechanism to speed up our CI jobs by building the images in "pull_request_target" workflow and pushing them to GitHub registry. That however had several drawbacks:

* CI image was complex when it comes to layer setup (we had to pre- cache installed dependencies by installing them from branch tip

* The pull_request_target is a very dangerous workflow, we had a number of security problems with it (and it's difficult to debug)

* Caching of `pip` and `uv` was not used because it increased size of the image significantly

This PR significantly improves the caching mechanisms for the images building of several advacements that were not possible before:

* The upload-artifacts@v4 action and improved stash action developed by @assignUser and published in "apache/infrastructure-actions" allows us to store all images (8GB per run) in artifacts rather than in registry - so we can do the image build once and share it with all the jobs.

* The uv speed is "enough" to allow occasional installation of Airlfow locally. This allows to utilize cache-mount and locally build uv cache, rather than rely on "remote" cache when we are building local images for breeze. The first time you build local breeze image it will take 2-5 more minutes (depending on your network speed, but because we can utilise cache mounts, every subsequent build should be very fast - even if all dependencies change. Using uv also allows to "always" reinstall airflow when you build the image even if single source file changed, because with cache it takes sub-seconds to reinstall airflow and all dependencies.

* the cache mounts are not included in the image size, and since we can export and import images in CI in artifacts and we do not need to rebuild them, the images shared as compressed artifacts are relatively small (2GB) - cache of `uv` is around 4GB on top of that so sharing image built in the "build image" job with other jobs in the same workflow is fast.

* we are still using registry cache for the "non-python" parts of the image - both CI and breeze image build speed benefit from using the image cache for system dependencies, database clients etc.

* documentation has been updated to reflect the new CI setup. The  diagrams showing the workflows of ours are no longer needed as the workflows are quite straightforward when they are looked at.

Fixes: #42999
Fixes: #43268

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: .github/actions/prepare_breeze_and_image/action.yml [hunk 1]
```diff
@@ -35,11 +35,23 @@ runs:
...
+      run: ./scripts/ci/import_docker_image.sh
       shell: bash
-      run: breeze ci-image pull --tag-as-latest
       if: inputs.pull-image-type == 'CI'
-    - name: Pull PROD image ${{ env.PYTHON_MAJOR_MINOR_VERSION }}:${{ env.IMAGE_TAG }}
+    - name: Restore PROD image ${{ env.PYTHON_MAJOR_MINOR_VERSION }}
+      uses: actions/download-artifact@v4
+      with:
+        name: "prod-image-docker-dump-${{ inputs.platform }}-${{ matrix.python-version }}"
+        path: "/tmp/docker/image.tar"
```
Comment (potiuk) [inline .github/actions/prepare_breeze_and_image/action.yml]:
```suggestion
        path: "/tmp/docker-image.tar"
```

File: .github/workflows/ci-image-build.yml [hunk 2]
```diff
@@ -189,7 +159,17 @@ ${{ inputs.do-build == 'true' && inputs.image-tag || '' }}"
...
+      - name: "Export docker storage"
+        run: >
+          ./scripts/ci/export_docker_image.sh
+          "ghcr.io/apache/airflow/main/ci/python${{ matrix.python-version }}"
+        if: inputs.upload-image-artifact == 'true'
+      - name: "Upload artifact"
+        uses: actions/upload-artifact@v4
+        with:
+          name: "ci-image-docker-dump-${{ inputs.platform }}-${{ matrix.python-version }}"
+          path: "/tmp/docker/image.tar"
```
Comment (potiuk) [inline .github/workflows/ci-image-build.yml]:
```suggestion
          path: "/tmp/docker-image.tar"
```

File: .github/actions/prepare_breeze_and_image/action.yml (hunk 1 above)
Comment (potiuk) [inline .github/actions/prepare_breeze_and_image/action.yml]:
```suggestion
        path: "/tmp/docker-image.tar"
```

File: .github/actions/prepare_breeze_and_image/action.yml [hunk 3]
```diff
@@ -35,11 +35,23 @@ runs:
     - name: Login to ghcr.io
       shell: bash
       run: echo "${{ env.GITHUB_TOKEN }}" | docker login ghcr.io -u ${{ github.actor }} --password-stdin
-    - name: Pull CI image ${{ env.PYTHON_MAJOR_MINOR_VERSION }}:${{ env.IMAGE_TAG }}
+    - name: Download CI image ${{ env.PYTHON_MAJOR_MINOR_VERSION }}
+      uses: actions/download-artifact@v4
+      with:
+        name: "ci-image-docker-dump-${{ inputs.platform }}-${{ matrix.python-version }}"
+        path: "/tmp/docker/image.tar"
```
Comment (potiuk) [inline .github/actions/prepare_breeze_and_image/action.yml]:
```suggestion
        path: "/tmp/docker-image.tar"
```

**Comment by gopidesupavan** [comment]
Nice looking forward for this :)

File: .github/actions/prepare_all_ci_images/action.yml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  host-python-version:
+    description: Python version used in host
+    value: ${{ steps.breeze.outputs.host-python-version }}
+runs:
+  using: "composite"
+  steps:
+    - name: "Cleanup docker"
+      run: ./scripts/ci/cleanup_docker.sh
+      shell: bash
+    # TODO: Currently we cannot loop through the list of python versions and have dynamic list of
```
Comment (potiuk) [inline .github/actions/prepare_all_ci_images/action.yml:38]: cc: @assignUser -> will need an option to restore several keys in a single action for that. You might see some people from our team contributing to the stash action of yours :)

**Comment by potiuk** [comment]
Also opened a "canary" build from "apache" repo to check how caching will work #45266

I will eventually close that one and will mark "ready for review" the other one, as we cannot get this one to be green due to pull_request_target workflow failing.

**Comment by potiuk** [comment]
Merged in https://github.com/apache/airflow/pull/45266

**Pull request closed without merging**